src/middleware/builtin.ts

```diff
@@ -287,47 +293,34 @@ export function matchEventType(pattern: EventTypePattern): Middleware<SlackEvent
   };
 }
 
-// TODO: breaking change: why does this method have to be invoked as a function with no args, while other similar
-// method like the `only*` ones do not require that? should make this consistent.
 /**
  * Filters out any event originating from the handling app.
  */
-export function ignoreSelf(): Middleware<AnyMiddlewareArgs> {
-  return async (args) => {
-    const botId = args.context.botId as string;
-    const botUserId = args.context.botUserId !== undefined ? (args.context.botUserId as string) : undefined;
-
-    if (isEventArgs(args)) {
-      if (args.event.type === 'message') {
-        // Once we've narrowed the type down to SlackEventMiddlewareArgs, there's no way to further narrow it down to
-        // SlackEventMiddlewareArgs<'message'> without a cast, so the following couple lines do that.
-        // TODO: there must be a better way; generics-based types for event and middleware arguments likely the issue
-        // should instead use a discriminated union
-        const message = args.message as unknown as SlackEventMiddlewareArgs<'message'>['message'];
-        if (message !== undefined) {
-        // TODO: revisit this once we have all the message subtypes defined to see if we can do this better with
-        // type narrowing
-        // Look for an event that is identified as a bot message from the same bot ID as this app, and return to skip
-          if (message.subtype === 'bot_message' && message.bot_id === botId) {
-            return;
-          }
-        }
+export const ignoreSelf: Middleware<AnyMiddlewareArgs> = async (args) => {
+  const { botId, botUserId } = args.context;
+
+  if (isEventArgs(args)) {
+    if (isMessageEventArgs(args)) {
+      const { message } = args;
+      // Look for an event that is identified as a bot message from the same bot ID as this app, and return to skip
+      if (message.subtype === 'bot_message' && message.bot_id === botId) {
```

Comment:
The combination of cleaning up the middleware types and use of type predicates lets us clean up the type casting and simplify the logic in this method ❤️